src/main/java/net/fabricmc/loader/util/version/SemanticVersion.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.util.version;
+
+import net.fabricmc.loader.api.Version;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+public class SemanticVersion implements Comparable<SemanticVersion>, Version {
+	private final int[] components;
+	private final String prerelease;
+	private final String build;
+	private String friendlyName;
+
+	public SemanticVersion(String version, boolean storeX) throws VersionParsingException {
+		int buildDelimPos = version.indexOf('+');
+		if (buildDelimPos >= 0) {
+			build = version.substring(buildDelimPos + 1);
+			version = version.substring(0, buildDelimPos);
+		} else {
+			build = null;
+		}
+
+		int dashDelimPos = version.indexOf('-');
+		if (dashDelimPos >= 0) {
+			prerelease = version.substring(dashDelimPos + 1);
+			version = version.substring(0, dashDelimPos);
+		} else {
+			prerelease = null;
+		}
+
+		if (version.endsWith(".")) {
+			throw new VersionParsingException("Negative version number component found!");
+		} else if (version.startsWith(".")) {
+			throw new VersionParsingException("Missing version component!");
+		}
+
+		String[] componentStrings = version.split("\\.");
+		if (componentStrings.length < 1) {
+			throw new VersionParsingException("Did not provide version numbers!");
+		}
+		components = new int[componentStrings.length];
+		for (int i = 0; i < componentStrings.length; i++) {
+			if (storeX && componentStrings[i].equals("x")) {
```

Comment:
I am pretty certain that usually `"x"`, `"X"` and `"*"` can be used for x ranges. Right now you only accept `"x"` and usually just return true if a `"*"` is used. Is there any reason why you are doing this?

Reply:
>Any of X, x, or * may be used to “stand in” for one of the numeric values in the [major, minor, patch] tuple.

As much is true, and it's decently easy to handle. The "*" always returning true is just a quick "passthrough".

Reply:
Has the next commit adequately solved this?

Reply:
yes, thank you